packages/file_selector/file_selector/test/file_selector_test.dart

```diff
@@ -0,0 +1,91 @@
+// Copyright 2020 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter_test/flutter_test.dart';
+import 'package:mockito/mockito.dart';
+import 'package:plugin_platform_interface/plugin_platform_interface.dart';
+import 'package:file_selector/file_selector.dart';
+import 'package:file_selector_platform_interface/file_selector_platform_interface.dart';
+
+void main() {
+  final mock = MockFileSelector();
+  FileSelectorPlatform.instance = mock;
+
+  test('getSavePath', () async {
+    final expectedPath = '/example/path';
+
+    final typeGroup = XTypeGroup(label: 'group', extensions: ['jpg', 'png']);
+
+    when(mock.getSavePath(
+      acceptedTypeGroups: [typeGroup],
+      initialDirectory: 'dir',
+      suggestedName: 'name',
+      confirmButtonText: 'save',
+    )).thenAnswer((_) => Future.value(expectedPath));
+
+    final result = await getSavePath(
+        acceptedTypeGroups: [typeGroup],
+        initialDirectory: 'dir',
+        suggestedName: 'name',
+        confirmButtonText: 'save');
+
+    expect(result, expectedPath);
+  });
+
+  test('getDirectoryPath', () async {
+    final expectedPath = '/example/path';
+
+    when(mock.getDirectoryPath(
+      initialDirectory: 'dir',
+      confirmButtonText: 'save',
+    )).thenAnswer((_) => Future.value(expectedPath));
+
+    final result = await getDirectoryPath(
+        initialDirectory: 'dir', confirmButtonText: 'save');
+
+    expect(result, expectedPath);
+  });
+
+  test('openFile', () async {
+    final file = XFile('path');
+
+    final typeGroup = XTypeGroup(label: 'group', extensions: ['jpg', 'png']);
+
+    when(mock.openFile(
+      acceptedTypeGroups: [typeGroup],
+      initialDirectory: 'dir',
+      confirmButtonText: 'load',
+    )).thenAnswer((_) => Future.value(file));
+
+    final result = await openFile(
+        acceptedTypeGroups: [typeGroup],
+        initialDirectory: 'dir',
+        confirmButtonText: 'load');
+
+    expect(result, isNotNull);
+  });
+
+  test('openFiles', () async {
+    final file = XFile('path');
+
+    final typeGroup = XTypeGroup(label: 'group', extensions: ['jpg', 'png']);
+
+    when(mock.openFiles(
+      acceptedTypeGroups: [typeGroup],
+      initialDirectory: 'dir',
+      confirmButtonText: 'load',
+    )).thenAnswer((_) => Future.value([file]));
+
+    final result = await openFiles(
+        acceptedTypeGroups: [typeGroup],
+        initialDirectory: 'dir',
+        confirmButtonText: 'load');
+
+    expect(result, isNotNull);
+  });
+}
```

Comment:
I might be missing something but my impression is that the platform implementations should test all those behaviours as at this particular moment the API is only receiving arguments and calling the platform interface with those.

Reply:
The platform implementations should test the API surface of the platform implementations. This plugin's test should be providing and end-to-end test, including the API that everyone will actually be using.

Reply:
Sorry, I was thinking of integration tests when writing that last comment. For unit tests:

>  the API is only receiving arguments and calling the platform interface with those.

How do you know that the API is doing that correctly without tests?

Reply:
Hi @stuartmorgan, no problem at all :)!

> How do you know that the API is doing that correctly without tests?

Because we are mocking the platform interface and making it return something only when it is called with specific arguments. 

For instance, on this test, we indicate the mock to return `[file]` only if it is called with those arguments.
```dart
    when(mock.openFiles(
      acceptedTypeGroups: [typeGroup],
      initialDirectory: 'dir',
      confirmButtonText: 'load',
    )).thenAnswer((_) => Future.value([file]));
```
And then we call the API and make sure it returns `[file]`. 
```dart
    final result = await openFiles(
        acceptedTypeGroups: [typeGroup],
        initialDirectory: 'dir',
        confirmButtonText: 'load');

    expect(result, [file]);
```
Currently the only case where this test would pass is if the API returns the result of calling platform interface with the same arguments.

Reply:
>> How do you know that the API is doing that correctly **without tests?**
>
> [...]
> For instance, **on this test**

That's a test, so isn't addressing my question.

This comment thread started because I was pointing out that there are many cases not covered by your unit tests (e.g., not providing the optional arguments). Your argument was that you don't need to test those cases because they are just calling through. I'm not asking how unit tests work, I'm asking how you can assert that the cases that you are *not* testing are, in fact, calling through correctly?

Let me put this a different way: which test will fail if someone adds code to `openFile` that attempts to do some preprocessing on, say, `initialDirectory`, without null-checking it?

Reply:
Hi Stuart, I expanded the tests coverage. PTAL